### ISSUE TITLE
Service not stopping (v1.2-9) from tray icon menu

### ISSUE THREAD

**Issue by RedBearAK** [rbreaves/kinto#504]
On Ubuntu 21.04 so using `systemd`, but the tray icon menu doesn't seem to be properly stopping the service anymore. It will stop, but then restart, as if I used the "Restart" option. 

The "Stop" option in the GUI window menu does work. So probably just need to slipstream in some change you made to the GUI code into the tray app. 

Also looks like if service is stopped via the GUI, the "Restart" option in the tray icon menu won't restart the service anymore.

**Comment by rbreaves** [comment]
I really could have sworn that I had fixed the system tray already, but apparently not. I must have wiped out some of my final fixes to that the night prior before making my final commit. I guess I didn't retest the restart and stop functionality after I saw it working.. copied over kinto-gui.py probably and not kintotray.py. Those 2 files have a lot of overlap but are different files and need to be.. so it kinda makes sense that in my hasty review I saw the changes I wanted to see without realizing they weren't duplicated, being shown twice mostly, as they should have been.

I pretty well know that is what happened because check_output got removed from the top but not later in the code for the sysv variable. Could have only happened that way if I forgot to copy over the file in its entirety as I originally had it yesterday. A stupid mistake from me trying to power through it and not have CI (Continuous Integration) checks in place on any commits in the dev & master branch.
 
Please test out the latest commit in master and I will then close this ticket if you find that it fully addresses the issue.

**Comment by rbreaves** [comment]
I really do think I might out to just put a freeze on PRs until I knock out the CI component for dev, master and submitted PRs. Instead of overwhelming myself with creating a CI workflow for multiple distros I think I will focus on making it rock solid for a single distro, Ubuntu Budgie. It will not catch edge cases, but something as silly as this messed up commit would have been caught early on without any manual testing as it'd have been pretty basic. Click sys tray -> click stop. Does xkeysnail stop? Yes/No - that'd be a fairly basic test that could probably be done in a fairly natural language tbh.

Not entirely sure how best to generate some of the tests, my initial thought is that it needs to be UI driven mostly - but I'll see. I guess the 2 types of tests will either be a more typical internal code based unit test vs UI automation driven tests.

Lettuce seems interesting, a lot like Capybara & Cucumber in the Ruby realm but Lettuce is for Python. https://codefellows.github.io/sea-python-401d2/lectures/lettuce.html

**Comment by RedBearAK** [comment]
I saw some of the redundancy between them when I was puzzling out how to do that new tweak option. Seems like a good opportunity to import certain functions from an external file. Aren't the stop/restart functions identical between them? 

The CI stuff is desirable I'm sure but won't it be complex to set up? It would be a shame to freeze all the PRs. Maybe just freeze changes to files other than `kinto.py`? 

This is my new daily driver and I don't have Boxes fully set up just yet, so I'm thinking the simplest way to test is just to copy over the `kintotray.py` file and run it.

**Comment by rbreaves** [comment]
I think I’ll implement a light version of unit tests & UI testing & CI before accepting further PRs. It will not be super complicated or thorough initially. I just need SOME basic & critical tests configured so that it can be built out over time.

**Comment by RedBearAK** [comment]
Just saved the tray file into place straight from the web browser and marked it executable to test. 

Enabling tray icon from GUI: Works. 

Disabling tray icon from GUI: Works, but check mark is still enabled in the menu when tray icon is gone. 

Disabling tray icon from GUI and then quitting GUI, when I restart GUI and re-enable tray icon, there is no check mark in the GUI menu even though tray icon is active. 

So there is some kind of disconnect in both directions for the check mark not mirroring the actual state of the tray icon. Cosmetic and minor, but I don't recall seeing that before. Not sure if there is any way I could have caused it by the way I installed the file, since it's working otherwise. 

Stopping from GUI menu and restarting from tray icon: Works. Solved. 

Restarting from tray icon: Works. 

Stopping from tray icon: Works. Stays stopped. That was the main issue. 

Seems like I'm not seeing a certain longer delay in the restart process that I thought I was seeing before. But the icon usually does cycle white-black-white when restarting. 

As far as I know the main function is fixed. 

>I think I’ll implement a light version of unit tests & UI testing & CI before accepting further PRs. It will not be super complicated or thorough initially. I just need SOME basic & critical tests configured so that it can be built out over time.

"Light" sounds good. I'd really like to get my "keypad and media key" fixes added if possible, as they solve a major irritation with all shortcuts involving the numeric keypad, and my strange "media" arrow keys. Copying and pasting the two large blocks by hand into any new config file is a pain.

**Comment by RedBearAK** [comment]
More good news, even if obvious: The ugly deprecation warnings are gone when running the tray file from terminal. I should have tested for that and reminded you before the release if it was still there. I remember I was still seeing the warnings late in the process, but I figured you were just waiting until you were done to slide the new changes into the tray app, which of course you thought you did, and therefore I felt bringing it up would have been redundant. Sorry about that. My nose for trouble failed me at the time.

**Comment by rbreaves** [comment]
It's all good, on the warning it was yea, a regression from the work I had actually already performed locally that simply missed the commit due to me not paying attention apparently. It was pretty apparent when I went back to run the tray directly in the terminal though lol. Python really wants things to be very explicit, so explicit it is - added the label param it wanted.

https://github.com/rbreaves/kinto/commit/ef60bf8a59e1f8dc728ab5f0f19af6f343773837
